go/ethadapter/blockprovider.go

```diff
@@ -0,0 +1,227 @@
+func (e *EthBlockProvider) streamBlocks(fromHeight *big.Int) {
+	atomic.StoreInt32(e.runningStatus, int32(statusCodeRunning))
+	for !e.stopped() {
+		// this will block if we're up-to-date with live blocks
+		block, err := e.nextBlockToStream(fromHeight)
+		if err != nil {
+			e.logger.Error("unexpected error while preparing block to stream, will retry in 1 sec", log.ErrKey, err)
+			time.Sleep(time.Second)
+			continue
+		}
+		e.logger.Trace("blockProvider streaming block", "height", block.Number(), "hash", block.Hash())
+		e.streamCh <- block // we block here until consumer takes it
+		// update stream state
+		e.currentHead = block.Header()
+	}
+}
+
+func (e *EthBlockProvider) nextBlockToStream(fromHeight *big.Int) (*types.Block, error) {
+	if e.currentHead == nil {
+		blk, err := e.ethClient.BlockByNumber(fromHeight)
+		if err == nil {
+			return blk, nil
+		}
+	}
+
+	head, err := e.awaitNewBlock()
+	if err != nil {
+		return nil, fmt.Errorf("no new block found from stream - %w", err)
+	}
+
+	// most common path should be: new head block that arrived is the next block, and needs to be sent
+	if head.ParentHash == e.currentHead.Hash() {
+		blk, err := e.ethClient.BlockByHash(head.Hash())
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch block with hash=%s - %w", head.Hash(), err)
+		}
+		return blk, nil
+	}
+
+	// and if not then, we find the latest canonical block we sent and try one after that
+	latestCanon, err := e.latestCanonAncestor(e.currentHead.Hash())
+	if err != nil {
+		return nil, fmt.Errorf("could not find ancestor on canonical chain for hash=%s - %w", e.currentHead.Hash(), err)
+	}
+	// and send the cannon block after the last sent (this may be a fork, or it may just be the next on the same branch)
+	blk, err := e.ethClient.BlockByNumber(increment(latestCanon.Number()))
+	if err != nil {
+		return nil, fmt.Errorf("could not find block after canon fork branch, height=%s - %w", increment(latestCanon.Number()), err)
+	}
+	return blk, nil
+}
+
+// stopped checks the status for stopped code
+func (e *EthBlockProvider) stopped() bool {
+	return atomic.LoadInt32(e.runningStatus) == int32(statusCodeStopped)
+}
+
+func (e *EthBlockProvider) latestCanonAncestor(blkHash gethcommon.Hash) (*types.Block, error) {
+	blk, err := e.ethClient.BlockByHash(blkHash)
+	if err != nil {
+		return nil, err
+	}
+	canonAtSameHeight, err := e.ethClient.BlockByNumber(blk.Number())
+	if err != nil {
+		return nil, err
+	}
+	if blk.Hash() != canonAtSameHeight.Hash() {
+		return e.latestCanonAncestor(blk.ParentHash())
+	}
+	return blk, nil
+}
+
+// awaitNewBlock takes a hash, it will block until it can return a head block with a different hash or error
+// (note: this can currently only be used by one caller at a time - not an issue for current usage)
+func (e *EthBlockProvider) awaitNewBlock() (*types.Header, error) {
+	// first we check if we're up-to-date
+	l1Head, f := e.ethClient.FetchHeadBlock()
```

Comment:
Yeah fair enough, have reworked so that awaitNewBlock just does the waiting for a new block. The caller checks if the provider is up-to-date and this method only gets called if it is up-to-date.